File: Backend/app/schemas/log.py
from pydantic import BaseModel, Field, AliasChoices
from datetime import datetime
from typing import Any, Dict, Optional

class LogBase(BaseModel):
  level: str
  message: str
  timestamp: str
  deviceid: str = Field(..., serialization_alias="deviceid", validation_alias=AliasChoices("deviceid", "deviceId", "sessionId"))
  url: Optional[str] = None
  stackTrace: Optional[str] = Field(None, serialization_alias="stackTrace", validation_alias="stackTrace")

  model_config = {
    "populate_by_name": True,
    "from_attributes": True
  }

class LogEventSchema(BaseModel):
  deviceId: str = Field(..., serialization_alias="deviceId", validation_alias="deviceId")
  sessionId: str = Field(..., serialization_alias="sessionId", validation_alias="sessionId")
  sessionStartedAt: Optional[str] = Field(None, serialization_alias="sessionStartedAt", validation_alias="sessionStartedAt")
  level: str
  message: str
  timestamp: str
  browser: str
  browserVersion: Optional[str] = Field(None, serialization_alias="browserVersion", validation_alias="browserVersion")
  deviceName: Optional[str] = Field(None, serialization_alias="deviceName", validation_alias="deviceName")
  os: Optional[str] = None
  latitude: Optional[float] = None
  longitude: Optional[float] = None
  url: str

  model_config = {
    "populate_by_name": True,
    "from_attributes": True
  }

class LogResponse(BaseModel):
  id: str
  deviceId: str = Field(..., serialization_alias="deviceId")
  level: str
  message: str
  timestamp: datetime
  url: Optional[str] = None
  stackTrace: Optional[str] = Field(None, serialization_alias="stackTrace")
  deviceid: str = Field(..., serialization_alias="deviceid")
  created_at: datetime
  location: Optional[Dict[str, Any]] = None
  
  # Flat fields mapped from Device
  browser: Optional[str] = None
  browserVersion: Optional[str] = Field(None, serialization_alias="browserVersion")
  deviceName: Optional[str] = Field(None, serialization_alias="deviceName")
  os: Optional[str] = None
  latitude: Optional[float] = None
  longitude: Optional[float] = None
  sessionStartedAt: Optional[datetime] = Field(None, serialization_alias="sessionStartedAt")

  model_config = {
    "populate_by_name": True,
    "from_attributes": True
  }

File: Frontend/src/types/log.ts
export interface LogEvent {
  id: string;
  level: "info" | "warn" | "error" | "debug";
  message: string;
  timestamp: string;
  deviceid: string;
  ip?: string;
  country?: string;
  city?: string;
  region?: string;
  latitude?: number;
  longitude?: number;
  sdkVersion?: string;
  appVersion?: string;
  userAgent?: string;
  url?: string;
  stackTrace?: string;
  location?: {
    href?: string;
    origin?: string;
    protocol?: string;
    host?: string;
    hostname?: string;
    port?: string;
    pathname?: string;
    search?: string;
    hash?: string;
  };
  
  // New SDK fields
  browser?: string;
  browserVersion?: string;
  deviceName?: string;
  os?: string;
  sessionStartedAt?: string;
}

export interface SessionInfo {
  deviceid: string;
  userAgent: string;
  url: string;
  duration: number; // In seconds
  logCount: number;
  errorCount: number;
  country?: string;
  city?: string;
  os?: string;
  browser?: string;
  
  // New SDK fields
  browserVersion?: string;
  deviceName?: string;
  latitude?: number;
  longitude?: number;
  sessionStartedAt?: string;
}
